### DISCUSSION TITLE
How to query nullable fields?

### DISCUSSION THREAD

**Discussion by Sheraff** [prisma/prisma#16056] [Q&A] [answered]
I have many different cases where I need to "query based on a nullable field being null" (or being "not null") and they seem to all require a different syntax, I'm a little lost and I've been reading the docs quite a lot without finding my answer.

As far as I can tell (from the docs and the TS autocompletion), the syntax of the query should depend on:
- the type of the relation: one-to-one, one-to-many, many-to-many
- whether we're querying the "bearer" (defines the `@relation(...)`) or the "target" (simple `@relation` attribute)
- the type of operation: `findFirst`, `findMany`, `update`, `updateMany`

## case 1: `update` the "bearer" of a one-to-one relation
With the following model, I want to **update a unique `SpotifyArtist` but only if their related `Artist` is not set**. What would be the correct query syntax?
```prisma
model Artist {
  id          String         @id @default(cuid())
  spotify     SpotifyArtist? @relation
}
model SpotifyArtist {
  id        String   @id
  artistId  String?  @unique
  artist    Artist?  @relation(fields: [artistId], references: [id])
}
```
```js
const spotifyId = 'aaa'
const artistId = 'bbb'
await prisma.spotifyArtist.update({
  where: {
    id: spotifyId,
    artistId: undefined
  },
  data: {
    artist: {
      connect: {
        id: artistId,
      }
    }
  }
})
```

## case 2: `findFirst` the "bearer" of a one-to-many relation
With the following model, I want to find **an `Album` with the name "foo" and whose related `Artist` is not set**. What would be the correct syntax?
```prisma
model Album {
  id          String   @id @default(cuid())
  name        String
  artistId    String?
  artist      Artist?  @relation(fields: [artistId], references: [id])

  @@unique([name, artistId])
}
model Artist {
  id         String   @id @default(cuid())
  albums     Album[]  @relation
}
```
```js
const match = 'aaa'
prisma.album.findFirst({
  where: {
    name: match,
    artist: null,
  }
})
```

## case 3: `findMany` the "target" of a one-to-one relation
With the following model, I want to find **all `Track` entries with no associated `File`**. What would be the correct syntax?
```prisma
model File {
  id        String   @id @default(cuid())
  trackId   String   @unique
  track     Track    @relation(fields: [trackId], references: [id], onDelete: Cascade)
}

model Track {
  id          String   @id @default(cuid())
  file        File?    @relation
}
```
```js
prisma.track.findMany({
  where: {file: {is: null}},
})
```

## case 4: `findMany` the "target" of a one-to-one relation
With the following model, I want to find **all `Album` entries with no associated `SpotifyAlbum`**. What would be the correct syntax?
```prisma
model Album {
  id          String        @id @default(cuid())
  name        String
  spotify     SpotifyAlbum? @relation
}
model SpotifyAlbum {
  id      String  @id
  albumId String? @unique
  album   Album?  @relation(fields: [albumId], references: [id])
}
```
```js
prisma.album.findMany({
  where: {
    name: albumName,
    spotify: null,
  }
})
```

**Comment by pantharshit00** [comment, marked as answer]
Hello @Sheraff,

 You have done all the queries correctly expect for the first one. 

> case 1: update the "bearer" of a one-to-one relation

You will need to use the `extendedWhereUnique` preview feature here. Read more about it here: https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#filter-on-non-unique-fields-with-userwhereuniqueinput. It will allow you to filter via `artist` in the update call.

```ts
  const spotifyId = "aaa";
  const artistId = "bbb";
  const data = await prisma.spotifyArtist.update({
    where: {
      id: spotifyId,
      artist: null
    },
    data: {
      artist: {
        connect: {
          id: artistId,
        },
      },
    },
  });
```


In all the other ones, the query you have provided is the correct query. 

I will recommend you to read the following docs section: https://www.prisma.io/docs/concepts/components/prisma-client/null-and-undefined. It details about usage of `null` vs `undefined`. `null` means the database value of `null` and `undefined` mean no value at all which will produce the same result as you would see when you will omit the field completely from the query.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
